org.eclipse.xtext.builder.tests/src/org/eclipse/xtext/builder/impl/IntegrationTest.java

```diff
@@ -317,6 +314,84 @@ protected void createTwoFilesInTwoReferencedProjects() throws Exception {
 		assertEquals(0, countMarkers(foo_file));
 		assertEquals(0, countMarkers(bar_file));
 	}
+	
+	@SuppressWarnings("restriction")
+	@Test public void testBuildOrderIsCorrect() throws Exception {
+		foo_project = createJavaProjectWithRootSrc("foo");
+		bar_project = createJavaProjectWithRootSrc("bar");
+
+		org.eclipse.core.internal.resources.Workspace workspace =
+				(org.eclipse.core.internal.resources.Workspace) ResourcesPlugin.getWorkspace();
+		IBuildConfiguration[] buildOrder = workspace.getBuildOrder();
+		assertEquals(bar_project.getProject(), buildOrder[0].getProject());
+		assertEquals(foo_project.getProject(), buildOrder[1].getProject());
+		addProjectReference(bar_project, foo_project);
+		
+		buildOrder = workspace.getBuildOrder();
+		assertEquals(foo_project.getProject(), buildOrder[0].getProject());
+		assertEquals(bar_project.getProject(), buildOrder[1].getProject());
+	}
+	
+	@SuppressWarnings("restriction")
+	@Test public void testBuildOrderIsWrong() throws Exception {
+		foo_project = createJavaProjectWithRootSrc("foo");
+		bar_project = createJavaProjectWithRootSrc("bar");
+
+		org.eclipse.core.internal.resources.Workspace workspace =
+				(org.eclipse.core.internal.resources.Workspace) ResourcesPlugin.getWorkspace();
+		IBuildConfiguration[] buildOrder = workspace.getBuildOrder();
+		assertEquals(bar_project.getProject(), buildOrder[0].getProject());
+		assertEquals(foo_project.getProject(), buildOrder[1].getProject());
+		JavaProjectSetupUtil.addProjectReference(bar_project, foo_project);
```

Comment:
having a comment that indicates that this sets a classpath ref would be nice

Reply:
Done